rust/worker/src/execution/operators/knn_merge.rs

```diff
@@ -0,0 +1,118 @@
+use chroma_error::{ChromaError, ErrorCodes};
+use thiserror::Error;
+use tonic::async_trait;
+
+use crate::execution::operator::Operator;
+
+use super::{fetch_log::FetchLogOutput, fetch_segment::FetchSegmentOutput, knn::Distance};
+
+#[derive(Clone, Debug)]
+pub struct KnnMergeOperator {
+    pub fetch: u32,
+}
+
+#[derive(Clone, Debug, Default)]
+pub struct PreKnnMergeState {
+    pub logs: Option<FetchLogOutput>,
+    pub segments: Option<FetchSegmentOutput>,
+    pub log_distance: Option<Vec<Distance>>,
+    pub segment_distance: Option<Vec<Distance>>,
+}
+
+#[derive(Debug)]
+pub struct KnnMergeInput {
+    pub logs: FetchLogOutput,
+    pub segments: FetchSegmentOutput,
+    pub log_distance: Vec<Distance>,
+    pub segment_distance: Vec<Distance>,
+}
+
+#[derive(Debug)]
+pub struct KnnMergeOutput {
+    pub logs: FetchLogOutput,
+    pub segments: FetchSegmentOutput,
+    pub distance: Vec<Distance>,
+}
+
+#[derive(Error, Debug)]
+pub enum KnnMergeError {
+    #[error("Error converting incomplete input")]
+    IncompleteInput,
+}
+
+impl ChromaError for KnnMergeError {
+    fn code(&self) -> ErrorCodes {
+        match self {
+            KnnMergeError::IncompleteInput => ErrorCodes::InvalidArgument,
+        }
+    }
+}
+
+impl TryFrom<PreKnnMergeState> for KnnMergeInput {
+    type Error = KnnMergeError;
+
+    fn try_from(value: PreKnnMergeState) -> Result<Self, KnnMergeError> {
+        if let PreKnnMergeState {
+            logs: Some(logs),
+            segments: Some(segments),
+            log_distance: Some(log_distance),
+            segment_distance: Some(segment_distance),
+        } = value
+        {
+            Ok(KnnMergeInput {
+                logs,
+                segments,
+                log_distance,
+                segment_distance,
+            })
+        } else {
+            Err(KnnMergeError::IncompleteInput)
+        }
+    }
+}
```

Comment:
unclear to me why this is needed; could this be handled by the caller?

Reply:
This could handled by the caller, but this is used by more than one place so I put it here for reuse. I'm considering about the migrating to the builder interface for this, which does basically the same thing but does not require manual handling like this.

Reply:
Update: This is deprecated, and the logic is put in to the orchestrator.